Mathlib/FieldTheory/SeparableDegree.lean

```diff
@@ -268,31 +249,539 @@ def embProdEmbOfIsAlgebraic [Algebra E K] [IsScalarTower F E K] (halg : Algebra.
+theorem natSepDegree_ne_zero (h : f.natDegree ≠ 0) : f.natSepDegree ≠ 0 := by
+  rw [natSepDegree, ne_eq, Finset.card_eq_zero, ← ne_eq, ← Finset.nonempty_iff_ne_empty]
+  use rootOfSplits _ (SplittingField.splits f) (ne_of_apply_ne _ h)
+  rw [Multiset.mem_toFinset, mem_aroots]
+  exact ⟨ne_of_apply_ne _ h, map_rootOfSplits _ (SplittingField.splits f) (ne_of_apply_ne _ h)⟩
+
+/-- A polynomial has zero separable degree if and only if it is constant. -/
+theorem natSepDegree_eq_zero_iff : f.natSepDegree = 0 ↔ f.natDegree = 0 :=
+  ⟨by simpa only [ne_eq, not_not] using mt (natSepDegree_ne_zero f), natSepDegree_eq_zero f⟩
+
+/-- A polynomial has non-zero separable degree if and only if it is non-constant. -/
+theorem natSepDegree_ne_zero_iff : f.natSepDegree ≠ 0 ↔ f.natDegree ≠ 0 :=
+  Iff.not <| natSepDegree_eq_zero_iff f
+
+/-- The separable degree of a non-zero polynomial is equal to its degree if and only if
+it is separable. -/
+theorem natSepDegree_eq_natDegree_iff (hf : f ≠ 0) :
+    f.natSepDegree = f.natDegree ↔ f.Separable := by
+  rw [natSepDegree, natDegree_eq_card_roots (SplittingField.splits f), aroots_def,
+    Multiset.toFinset_card_eq_card_iff_nodup,
+    ← separable_map (algebraMap F f.SplittingField)]
+  exact nodup_roots_iff_of_splits _ (map_ne_zero hf) <|
+    (splits_map_iff _ _).2 (SplittingField.splits f)
+
+/-- If a polynomial is separable, then its separable degree is equal to its degree. -/
+theorem natSepDegree_eq_natDegree_of_separable (h : f.Separable) :
+    f.natSepDegree = f.natDegree :=
+  (natSepDegree_eq_natDegree_iff f (fun h' ↦ not_separable_zero (h' ▸ h))).2 h
+
+/-- If a polynomial splits over `E`, then its separable degree is equal to
+the number of distinct roots of it over `E`. -/
+theorem natSepDegree_eq_of_splits (h : f.Splits (algebraMap F E)) :
+    f.natSepDegree = (f.aroots E).toFinset.card := by
+  let i := SplittingField.lift f h
+  have heq := (f.map (algebraMap F f.SplittingField)).map_roots_le_of_injective i.injective
+  rw [AlgHom.toRingHom_eq_coe, RingHom.coe_coe, map_map, AlgHom.comp_algebraMap_of_tower] at heq
+  replace heq := congr_arg Finset.card <| congr_arg Multiset.toFinset <|
+    Multiset.eq_of_le_of_card_le heq <| by rw [Multiset.card_map, ← natDegree_eq_card_roots h,
+      ← natDegree_eq_card_roots <| SplittingField.splits f]
+  have := Finset.card_image_of_injective
+    (f.map (algebraMap F f.SplittingField)).roots.toFinset i.injective
+  rw [AlgHom.toRingHom_eq_coe, RingHom.coe_coe] at this
+  rw [Multiset.toFinset_map, this] at heq
+  convert heq
+
+section
+
+variable (E)
+
+/-- The separable degree of a polynomial is equal to
+the number of distinct roots of it over any algebraically closed field. -/
+theorem natSepDegree_eq_of_isAlgClosed [IsAlgClosed E] :
+    f.natSepDegree = (f.aroots E).toFinset.card :=
+  natSepDegree_eq_of_splits f (IsAlgClosed.splits_codomain f)
+
+end
+
+@[simp]
+theorem natSepDegree_C_mul {x : F} (hx : x ≠ 0) :
```

Comment:
`f` should be explicit here and in some of the lemma below (not sure if it already is).

Reply:
In [Polynomial.natDegree_C_mul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Polynomial/Degree/Lemmas.html#Polynomial.natDegree_C_mul) the `f` is implicit.